Generate one log file per minikube command

bharathkkb opened the pull request:
<!-- 🎉 Thank you for contributing to minikube! 🎉 Here are some hints to get your PR merged faster:

1. Your PR title will be included in the release notes, choose it carefully
2. If the PR fixes an issue, add "fixes #<issue number>" to the description.
3. If the PR is a user interface change, please include a "before" and "after" example.
4. If the PR is a large design change, please include an enhancement proposal:
https://github.com/kubernetes/minikube/tree/master/enhancements
-->
fixes https://github.com/kubernetes/minikube/issues/9156 based on https://github.com/kubernetes/minikube/issues/9156#issuecomment-696195859

- By default if `log_file` is not specified, we write to `$TMPDIR/minikube_<argv[1]>_<hash>.log`
- if `log_dir` specified, we write to `$LOG_DIR/minikube_<argv[1]>_<hash>.log`
- Any binary runs with same args will result in logs being appended to the same logfile

### before
```sh
for i in {1..3}; do minikube status; done
```
creates
```
minikube.hostname.user.log.INFO.20210209-141816.44848
minikube.hostname.user.log.INFO.20210209-141818.44872
minikube.hostname.user.log.INFO.20210209-141819.44913
minikube.hostname.user.log.WARNING.20210209-141816.44848
minikube.hostname.user.log.WARNING.20210209-141818.44872
minikube.hostname.user.log.WARNING.20210209-141819.44913
```
each containing 
```
Log file created at: 2021/02/09 14:23:21
Running on machine: hostname
Binary: Built with gc go1.15.7 for darwin/amd64
Log line format: [IWEF]mmdd hh:mm:ss.uuuuuu threadid file:line] msg
```
### after
```sh
for i in {1..3}; do ./out/minikube status; done
```
creates
```
minikube_status_f744034489211de89319363bc95da8daa2870d09.log
```
containing
```
Log file created at: 2021/02/09 14:22:15
Running on machine: hostname
Binary: Built with gc go1.14.15 for darwin/amd64
Log line format: [IWEF]mmdd hh:mm:ss.uuuuuu threadid file:line] msg
Log file created at: 2021/02/09 14:22:16
Running on machine: hostname
Binary: Built with gc go1.14.15 for darwin/amd64
Log line format: [IWEF]mmdd hh:mm:ss.uuuuuu threadid file:line] msg
Log file created at: 2021/02/09 14:22:16
Running on machine: hostname
Binary: Built with gc go1.14.15 for darwin/amd64
Log line format: [IWEF]mmdd hh:mm:ss.uuuuuu threadid file:line] msg
```

medyagh reviewed (changes requested):
this is a good PR, thank @bharathkkb for fixing this.

we will need to update the documentation on the https://minikube.sigs.k8s.io/docs/handbook/troubleshooting/
since the command we suggest to find log files wont be correct anymore
```
find $TMPDIR -mtime -1 -type f -name "*minikube*INFO*" -ls  2>/dev/null
```
(or can we make it in a way that the same suggestion on the site still works?)
I had to change it to not have the *INFO
```
 find $TMPDIR -mtime -1 -type f -name "*minikube*" -ls  2>/dev/null 
```
and I confirmed your fix only creates one file per command.
is it possible that we set a limit on how big this file can be ? maybe do not allow it to be bigger than 1 mb ?

```
medya@~/workspace/minikube (fix-multiple-log-files) $ cat /var/folders/s8/wxxccj3x7mncysv_zzm5w_r400h78j/T//minikube_version_a24f39ccb0b3c744e7472b70ec1372730a3ee53d.log
Log file created at: 2021/02/12 12:18:27
Running on machine: medya-macbookpro3
Binary: Built with gc go1.15.6 for darwin/amd64
Log line format: [IWEF]mmdd hh:mm:ss.uuuuuu threadid file:line] msg
I0212 12:18:27.500643   18787 out.go:229] Setting OutFile to fd 1 ...
I0212 12:18:27.500936   18787 out.go:281] isatty.IsTerminal(1) = true
I0212 12:18:27.500939   18787 out.go:242] Setting ErrFile to fd 2...
I0212 12:18:27.500942   18787 out.go:281] isatty.IsTerminal(2) = true
I0212 12:18:27.501025   18787 root.go:307] Updating PATH: /Users/medya/.minikube/bin
I0212 12:18:27.526953   18787 out.go:119] minikube version: v1.17.1
I0212 12:18:27.551752   18787 out.go:119] commit: a44e009c6d54971d208f2e2789fc8473b0bec789
Log file created at: 2021/02/12 12:18:57
Running on machine: medya-macbookpro3
Binary: Built with gc go1.15.6 for darwin/amd64
Log line format: [IWEF]mmdd hh:mm:ss.uuuuuu threadid file:line] msg
I0212 12:18:57.434975   18896 out.go:229] Setting OutFile to fd 1 ...
I0212 12:18:57.435169   18896 out.go:281] isatty.IsTerminal(1) = true
I0212 12:18:57.435172   18896 out.go:242] Setting ErrFile to fd 2...
I0212 12:18:57.435175   18896 out.go:281] isatty.IsTerminal(2) = true
I0212 12:18:57.435264   18896 root.go:307] Updating PATH: /Users/medya/.minikube/bin
I0212 12:18:57.461203   18896 out.go:119] minikube version: v1.17.1
I0212 12:18:57.485673   18896 out.go:119] commit: a44e009c6d54971d208f2e2789fc8473b0bec789
Log file created at: 2021/02/12 12:18:58
Running on machine: medya-macbookpro3
Binary: Built with gc go1.15.6 for darwin/amd64
Log line format: [IWEF]mmdd hh:mm:ss.uuuuuu threadid file:line] msg
I0212 12:18:58.228372   18911 out.go:229] Setting OutFile to fd 1 ...
I0212 12:18:58.228561   18911 out.go:281] isatty.IsTerminal(1) = true
I0212 12:18:58.228564   18911 out.go:242] Setting ErrFile to fd 2...
I0212 12:18:58.228567   18911 out.go:281] isatty.IsTerminal(2) = true
I0212 12:18:58.228650   18911 root.go:307] Updating PATH: /Users/medya/.minikube/bin
I0212 12:18:58.254501   18911 out.go:119] minikube version: v1.17.1
I0212 12:18:58.279179   18911 out.go:119] commit: a44e009c6d54971d208f2e2789fc8473b0bec789
Log file created at: 2021/02/12 12:19:23
Running on machine: medya-macbookpro3
Binary: Built with gc go1.15.6 for darwin/amd64
Log line format: [IWEF]mmdd hh:mm:ss.uuuuuu threadid file:line] msg
I0212 12:19:23.902215   18946 out.go:229] Setting OutFile to fd 1 ...
I0212 12:19:23.902501   18946 out.go:281] isatty.IsTerminal(1) = true
I0212 12:19:23.902503   18946 out.go:242] Setting ErrFile to fd 2...
I0212 12:19:23.902507   18946 out.go:281] isatty.IsTerminal(2) = true
I0212 12:19:23.902590   18946 root.go:307] Updating PATH: /Users/medya/.minikube/bin
I0212 12:19:23.928504   18946 out.go:119] minikube version: v1.17.1
I0212 12:19:23.953217   18946 out.go:119] commit: a44e009c6d54971d208f2e2789fc8473b0bec789
Log file created at: 2021/02/12 12:19:35
Running on machine: medya-macbookpro3
Binary: Built with gc go1.15.6 for darwin/amd64
Log line format: [IWEF]mmdd hh:mm:ss.uuuuuu threadid file:line] msg
I0212 12:19:35.165196   18989 out.go:229] Setting OutFile to fd 1 ...
I0212 12:19:35.165391   18989 out.go:281] isatty.IsTerminal(1) = true
I0212 12:19:35.165393   18989 out.go:242] Setting ErrFile to fd 2...
I0212 12:19:35.165397   18989 out.go:281] isatty.IsTerminal(2) = true
I0212 12:19:35.165479   18989 root.go:307] Updating PATH: /Users/medya/.minikube/bin
I0212 12:19:35.191324   18989 out.go:119] minikube version: v1.17.1
I0212 12:19:35.216018   18989 out.go:119] commit: a44e009c6d54971d208f2e2789fc8473b0bec789
Log file created at: 2021/02/12 12:19:35
Running on machine: medya-macbookpro3
Binary: Built with gc go1.15.6 for darwin/amd64
Log line format: [IWEF]mmdd hh:mm:ss.uuuuuu threadid file:line] msg
I0212 12:19:35.818268   19004 out.go:229] Setting OutFile to fd 1 ...
I0212 12:19:35.818565   19004 out.go:281] isatty.IsTerminal(1) = true
I0212 12:19:35.818568   19004 out.go:242] Setting ErrFile to fd 2...
I0212 12:19:35.818571   19004 out.go:281] isatty.IsTerminal(2) = true
I0212 12:19:35.818649   19004 root.go:307] Updating PATH: /Users/medya/.minikube/bin
I0212 12:19:35.844603   19004 out.go:119] minikube version: v1.17.1
I0212 12:19:35.869088   19004 out.go:119] commit: a44e009c6d54971d208f2e2789fc8473b0bec789
Log file created at: 2021/02/12 12:19:36
Running on machine: medya-macbookpro3
Binary: Built with gc go1.15.6 for darwin/amd64
Log line format: [IWEF]mmdd hh:mm:ss.uuuuuu threadid file:line] msg
I0212 12:19:36.483261   19019 out.go:229] Setting OutFile to fd 1 ...
I0212 12:19:36.483460   19019 out.go:281] isatty.IsTerminal(1) = true
I0212 12:19:36.483463   19019 out.go:242] Setting ErrFile to fd 2...
I0212 12:19:36.483466   19019 out.go:281] isatty.IsTerminal(2) = true
I0212 12:19:36.483561   19019 root.go:307] Updating PATH: /Users/medya/.minikube/bin
I0212 12:19:36.509380   19019 out.go:119] minikube version: v1.17.1
I0212 12:19:36.534035   19019 out.go:119] commit: a44e009c6d54971d208f2e2789fc8473b0bec789

```

medyagh commented:
I really like that every command has its own log file

```
medya@~/workspace/minikube (fix-multiple-log-files) $ find $TMPDIR -mtime -1 -type f -name "*minikube*" -ls  2>/dev/null 
41323719       16 -rw-r--r--    1 medya            primarygroup         5026 Feb 12 12:19 /var/folders/s8/wxxccj3x7mncysv_zzm5w_r400h78j/T//minikube_version_a24f39ccb0b3c744e7472b70ec1372730a3ee53d.log
41324222        8 -rw-r--r--    1 medya            primarygroup         1430 Feb 12 12:24 /var/folders/s8/wxxccj3x7mncysv_zzm5w_r400h78j/T//minikube_stop_71147ec8404058d0a5eaa842851c91f2c9e7fb78.log
```

prezha commented:
maybe we can have one log file per profile (limited with some max size)?

medyagh commented:
@bharathkkb I like this PR, would u like to address the TODO and send it for final review to be merged?

bharathkkb commented:
Hi @medyagh 
Just finishing up the tests and will have this up for review shortly.

medyagh commented:
CC @spowelljr  our new klog expert for a review

bharathkkb commented:
@medyagh
I have implemented a max size check so it will create a new log file for anything larger than 1M. Also added integration tests

spowelljr commented on cmd/minikube/main.go:
```diff
@@ -121,6 +124,39 @@ func (lb machineLogBridge) Write(b []byte) (n int, err error) {
 }
 
+// check if a file's size is greater than or equal to max size in KB
+func checkLogFileMaxSize(file string, maxSizeKB int64) bool {
```
Comments for funcs should start with the name of the func itself.

https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences

bharathkkb replied:
done

spowelljr commented on test/integration/error_log_test.go:
```diff
@@ -0,0 +1,124 @@
+				foundSmall := false
+				maxSize := 1024 * 1024 // 1M
+				for _, logFile := range logFiles {
+					if int(logFile.Size()) < maxSize && !foundSmall {
```
Refactoring `int(logFile.Size()) < maxSize` into a variable like `isSmall` would clean up the if/else blocks

spowelljr replied:
I just realized this is the test file, this isn't a huge deal.

bharathkkb replied:
good idea done

spowelljr commented on test/integration/error_log_test.go:
```diff
@@ -0,0 +1,124 @@
+		}
+		return nil
+	})
+	return logFiles, err
```
I don't think it really makes sense to return `err` in this func, it's not possible for `err` to be anything than `nil`, so we'll just doing an unnecessary error check.

spowelljr replied:
I just realized this is in the test file, this isn't a huge deal.

bharathkkb replied:
Couldn't there be a case where `filepath.Walk` returns an error in which case the `err` would not be nil? Sorry very much a newbie to golang so apologies if this is a stupid question.

spowelljr replied:
You are correct, I assumed that the only way to return an error would be from the func you defined as the second argument, but upon inspection of the source code of `filepath.Walk` there is a path where a non-nil error could be returned even if your func always returns `nil`.

spowelljr commented on test/integration/error_log_test.go:
```diff
@@ -0,0 +1,124 @@
+			for i := 0; i < test.runCount; i++ {
+				rr, err := Run(t, exec.CommandContext(ctx, Target(), args...))
+				if err != nil {
+					t.Errorf("failed to clean up: args %q: %v", rr.Command(), err)
```
Also, any point where the failure will affect the rest of the test, should do `t.Fatalf` so it doesn't continue to fail the rest of the checks and pollute the logs unnecessarily.

bharathkkb replied:
I have reformatted it to follow the other tests like https://github.com/kubernetes/minikube/blob/59cca690778b1981cbe2da3541971ff023d058a6/test/integration/addons_test.go#L70

spowelljr commented on site/content/en/docs/handbook/troubleshooting.md:
```diff
@@ -19,15 +19,15 @@ Example:
 ## Post-mortem minikube debug logs
 
-minikube stores post-mortem INFO logs in the temporary directory of your system. On macOS or Linux, it's easy to get a list of recent INFO logs:
+minikube stores post-mortem logs in the temporary directory of your system. One logfile is created per subcommand and any subsequent invocations of the subcommand with the same args will append to the same file. On macOS or Linux, it's easy to get a list of recent logs:
```
I think the `One log file per subcommand` is slightly misleading since once the file is over a MB a new file is created. Also, need a space between `logfile`.

bharathkkb replied:
Added a line to capture this behavior

priyawadhwa commented on test/integration/error_log_test.go:
```diff
@@ -0,0 +1,125 @@
+		{
+			command:          "start",
+			args:             append([]string{"--memory=2200", "--wait=true"}, StartArgs()...),
+			runCount:         20, // calling this 20 times should create 2 files with 1 greater than 1M
```
I'm concerned about running `minikube start` 20 times for this test -- this test will take minimum 20 minutes to run. 

Instead I would remove the `minikube start` subtest entirely and run `minikube status` 100 times (or whatever it takes) to make sure 2 log files are created.

bharathkkb replied:
@priyawadhwa did a quick test and for around 500 times x `minikube status` only generates 87K, so we may need quite a lot of status calls (~5K) to fill up 1M and create 2 files.

Another alternative might be to fill up the log with some mock data(<1M) and expect that minikube handles the creation of new log correctly on subsequent calls. However not sure if that would be a best practice since this is an integration test.

spowelljr replied:
@bharathkkb Maybe try `minikube logs` much shorter running time that start and generates quite a few logs.

medyagh replied:
@bharathkkb i really like that added integration test, I agree with @priyawadhwa lets try this test as a Sub Test for TestErrorSpam

we should not start a new cluster for this test, and it goes well with logging and spammming.
since we dont wanna spam TMPFolder

we can have a sub test that runs `minikube start --dry-run `

or minikube version

priyawadhwa reviewed (changes requested):
Thanks for submitting this fix! Left a comment around the integration test.

bharathkkb commented:
@medyagh I have switched this over to a subtest (let me know if this what you had in mind) and instead of `start`, I have used `logs` as suggested by @spowelljr

medyagh commented:
/ok-to-test

medyagh commented:
@bharathkkb in the test Error Spam I see this

```

=== CONT  TestErrorSpam/logs
error_spam_test.go:147: (dbg) Done: out/minikube-linux-amd64 logs -p nospam-20210303223533-1048494: (6.383692358s)
error_spam_test.go:161: failed to find expected number of log files: cmd logs: expected: 2 got 6
error_spam_test.go:172: expected to find only one file less than 1M: cmd logs:
error_spam_test.go:197: Cleaning up logfile /tmp/minikube_logs_c0e7076acab36180ef806327f1e14d12ce77a9ec_0.log ...
error_spam_test.go:197: Cleaning up logfile /tmp/minikube_logs_e62f2588220a8558239ba871be07e2c348f43e22_0.log ...
error_spam_test.go:197: Cleaning up logfile /tmp/minikube_logs_e62f2588220a8558239ba871be07e2c348f43e22_1.log ...
error_spam_test.go:197: Cleaning up logfile /tmp/minikube_logs_e62f2588220a8558239ba871be07e2c348f43e22_2.log ...
```


are we spamming the UI with stderr ?

medyagh commented:
@bharathkkb lets fix the test and then it looks good

medyagh commented:
hi @bharathkkb gentel reminder about this PR

spowelljr commented on test/integration/error_spam_test.go:
```diff
@@ -104,4 +107,103 @@ func TestErrorSpam(t *testing.T) {
+		{
+			command:          "start",
+			args:             []string{"--dry-run", "--log_dir", os.TempDir()},
+			runCount:         120, // calling this 15 times should create 2 files with 1 greater than 1M
```
Count in comment is outdated

bharathkkb replied:
good catch, done

spowelljr commented on test/integration/error_spam_test.go:
```diff
@@ -104,4 +107,103 @@ func TestErrorSpam(t *testing.T) {
+					if isSmall && !foundSmall {
+						foundSmall = true
+					} else if isSmall && foundSmall {
+						t.Errorf("expected to find only one file less than 1M: cmd %s:", test.command)
```
I think using `1 MB` is a more common abbreviation for megabyte

bharathkkb replied:
Done

bharathkkb commented:
@medyagh @spowelljr 
I have been trying to look into the failing tests in CI. Part of it was that the `logs` command was unpredictable w.r.t amount of content being written so I have switched to `minikube start --dry-run` as @medyagh  suggested. I also added a pre step so that any previous run logs are cleaned up and then the tests are run followed by asserting n number of logfiles found. 

However looking at the tests I still see that there are [cases](https://storage.googleapis.com/minikube-builds/logs/10425/7972124/Docker_Linux_crio.html#fail_TestErrorSpam%2fpause) where the logs could not be cleaned up causing failures. Are the tests being run in parallel on the same runner? Another workaround could be instead of just [using the cmd](https://github.com/kubernetes/minikube/blob/bd406f50a5700a9b6849f3b3c12ecd2d65be0bc6/test/integration/error_spam_test.go#L189) to match logfiles, we can [generate](https://github.com/kubernetes/minikube/blob/bd406f50a5700a9b6849f3b3c12ecd2d65be0bc6/cmd/minikube/main.go#L139) the expected SHA and use that to match instead.
